### PR TITLE
etcdserverpb: add "retry" flag to WatchResponse

### PR DESCRIPTION
**Pull request by lavacat**
Problem: client relies on string value of CancelReason to determine if watcher should be retried. This is too fragile. See https://github.com/etcd-io/etcd/discussions/14992

Solution: add explicit `retry` flag

fixes: https://github.com/etcd-io/etcd/issues/15058

Signed-off-by: Bogdan Kanivets <bkanivets@apple.com>

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: api/etcdserverpb/rpc.proto [hunk 1]
```diff
@@ -834,6 +834,9 @@ message WatchResponse {
   // framgment is true if large watch response was split over multiple responses.
   bool fragment = 7 [(versionpb.etcd_version_field)="3.4"];
 
+  // retry is true if watcher should be retried on cancel.
+  bool retry = 8 [(versionpb.etcd_version_field)="3.5"];
```
Comment (lavacat) [inline api/etcdserverpb/rpc.proto:838]: I'll backport to 3.5, so setting version 3.5 here

**Comment by lavacat** [comment]
This code is covered by TestV3AuthWatchAndTokenExpire integration test.
Looked at adding unit test to v3rpc/watch_test.go, but it requires setting up serverWatchStream.recvLoop.

File: api/etcdserverpb/rpc.proto (hunk 1 above)
Comment (serathius) [reply api/etcdserverpb/rpc.proto:838]: No backports of fields!

**Comment by serathius** [comment]
This is a small change but with a big impact on important API semantics. Please start with creating issue discussing problem you want to solve and why solution you propose is worth adopting.

**Review by serathius — CHANGES_REQUESTED** [review]
Would like to first discuss different options before picking a solution.

**Comment by lavacat** [comment]
@serathius do you mind continuing discussion in https://github.com/etcd-io/etcd/issues/15058? Here is my [comment](https://github.com/etcd-io/etcd/issues/15058#issuecomment-1399342701) explainig `retry`. Other folks commented about this, so don't want to loose context.

**Review by mitake — COMMENTED** [review]
After having consensus about the approach, I think we should update the changelog for describing the new field (or changes for other approach): https://github.com/etcd-io/etcd/blob/main/CHANGELOG/CHANGELOG-3.6.md

**Pull request closed without merging**